### PR TITLE
graphqlbackend: only allow SOAP users to delete other SOAP users

### PR DESCRIPTION
**Pull request by bobheadxi**
This change prevents non-SOAP users from deleting SOAP users.

This is not much of an issue yet because SOAP users are shortlived and unlikely to be deleted by a Cloud customer admin, but is relevant for https://github.com/sourcegraph/sourcegraph/pull/47059 and https://github.com/sourcegraph/sourcegraph/pull/47235 where we want to introduce the concept of a SOAP service account.

Stack fixes https://github.com/sourcegraph/customer/issues/1929

## Test plan

New and existing automated tests

### CODE REVIEW

**Comment by bobheadxi** [comment]
Current dependencies on/for this PR:
* main
  * **PR #47234** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/47234" target="_blank">Graphite</a>  👈
    * **PR #47059** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/47059" target="_blank">Graphite</a> 
      * **PR #47235** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/47235" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/47234?utm_source=stack-comment).

File: cmd/frontend/graphqlbackend/site_admin.go [hunk 1]
```diff
@@ -109,6 +109,9 @@ func (r *schemaResolver) DeleteUsers(ctx context.Context, args *struct {
 	if err != nil {
 		return nil, errors.Wrap(err, "list users by IDs")
 	}
+	if len(users) == 0 {
+		return nil, errors.Newf("no users found with IDs: %+v", ids)
+	}
```
Comment (bobheadxi) [inline cmd/frontend/graphqlbackend/site_admin.go:114]: This missing check causes the existing tests to not actually test user deletion

**Review by michaellzc — APPROVED** [review]
nice!

File: cmd/frontend/graphqlbackend/site_admin.go (hunk 1 above)
Comment (unknwon) [reply cmd/frontend/graphqlbackend/site_admin.go:114]: This `if` is important, however from a user's POV, if the user(s) intended to be deleted is/are already gone, why is that an error? (i.e. achieve the end result that I want anyway)

File: cmd/frontend/graphqlbackend/site_admin.go [hunk 2]
```diff
@@ -120,6 +123,19 @@ func (r *schemaResolver) DeleteUsers(ctx context.Context, args *struct {
 			return nil, errors.Wrap(err, "list external accounts")
 		}
 		for _, acct := range extAccounts {
+			// If the delete target is a SOAP user, make sure the actor is also a SOAP
+			// user - regular users should not be able to delete SOAP users.
+			if acct.ServiceType == auth.SourcegraphOperatorProviderType {
+				actorSoapAccts, err := r.db.UserExternalAccounts().List(ctx, database.ExternalAccountsListOptions{
+					UserID:      a.UID,
+					ServiceType: auth.SourcegraphOperatorProviderType,
+				})
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/site_admin.go:132]: https://github.com/sourcegraph/sourcegraph/blob/cb3f827cf85bf6a088e9e10e964a9456cd98b694/internal/actor/actor.go#L39-L40 😁

File: cmd/frontend/graphqlbackend/site_admin.go (hunk 1 above)
Comment (bobheadxi) [reply cmd/frontend/graphqlbackend/site_admin.go:114]: I think it is common to error in these cases, e.g. `git branch -D b1 b2 b3` errors for all 3 branches

Comment (unknwon) [reply cmd/frontend/graphqlbackend/site_admin.go:114]: If you look into how [`HardDeleteList`](https://github.com/sourcegraph/sourcegraph/blob/cb3f827cf85bf6a088e9e10e964a9456cd98b694/internal/database/users.go#L575) and [`DeleteList`](https://github.com/sourcegraph/sourcegraph/blob/cb3f827cf85bf6a088e9e10e964a9456cd98b694/internal/database/users.go#L518) are implemented, they do not care about non-exist users (i.e. `user_id IN (?)`), it's like `rm -rf`, as the user I really do not want to know which of them did not exist at all, just want all of them gone.

_This comment is non-blocking_

File: cmd/frontend/graphqlbackend/site_admin.go (hunk 2 above)
Comment (bobheadxi) [reply cmd/frontend/graphqlbackend/site_admin.go]: Updated to use this instead!

File: cmd/frontend/graphqlbackend/site_admin.go [hunk 3]
```diff
@@ -120,6 +130,15 @@ func (r *schemaResolver) DeleteUsers(ctx context.Context, args *struct {
 			return nil, errors.Wrap(err, "list external accounts")
 		}
 		for _, acct := range extAccounts {
+			// If the delete target is a SOAP user, make sure the actor is also a SOAP
+			// user - regular users should not be able to delete SOAP users.
+			if acct.ServiceType == auth.SourcegraphOperatorProviderType {
+				if !a.SourcegraphOperator {
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/site_admin.go:136]: LOL nit: if my eyes are working, we can make these two `if`s into one

```go
if acct.ServiceType == auth.SourcegraphOperatorProviderType && !a.SourcegraphOperator {
```

**Pull request merged**